### PR TITLE
FIX Cert length (Issue #4012)

### PR DESCRIPTION
**Pull request by mapedraza**
This PR is intended to help solving the Issue #4012

Those changes are the one received in skype from @kzangeli. Uploaded here as received.

Before merging this PR, needs to be completed:
- [x] Verify functionality
- [x] Add tests
- ~~Document (if necessary)~~
- [x] CHANGES_NEXT_RELEASE entry

### CODE REVIEW

File: src/app/contextBroker/contextBroker.cpp [hunk 1]
```diff
@@ -719,7 +721,9 @@ static int loadFile(char* path, char* out, int outSize)
     return -1;
```
Comment (kzangeli) [inline src/app/contextBroker/contextBroker.cpp:721]: [My fault] This line needs to return NULL and not -1.

There's one more "return -1" in the function that also (of course) needs the same change

Comment (mapedraza) [inline src/app/contextBroker/contextBroker.cpp:721]: Suggested by @kzangeli through Skype
```suggestion
    return NULL;
```

**Comment by fgalan** [comment]
CNR entry has to be included

With regards to .test, note this in httpsPrepare.sh (an utility script used by HTTPS tests)

```
openssl genrsa -out "$keyFileName" 2048 > /dev/null 2>&1
```

The 2048 sholuld be unhardwired to a httpsPrepare.sh parameter, then used to mantain the current test for 2048 and add new ones with a larger size (for instance the 2053 mentioned in issue #4012)

File: src/app/contextBroker/contextBroker.cpp [hunk 2]
```diff
@@ -1239,20 +1243,26 @@ int main(int argC, char* argV[])
 
   if (https)
   {
-    // FIXME P3: we suspect that loadFile() is not working well as it doesn't include the
-    // char terminator \0 at the end, thus causing valgrind errors. Using calloc will ensure
-    // that all the buffer is initialized with \0, thus avoiding the problem, although it would
-    // be better to fix loadFile()
-    // See issue https://github.com/telefonicaid/fiware-orion/issues/3925 for more detail
```
Comment (fgalan) [inline src/app/contextBroker/contextBroker.cpp:1246]: Review if valgrind pass is disabled in some .test due to this, re-enabling them again in that case.

**Comment by fgalan** [comment]
This PR was created by TEF @mapedraza on behalf the FIWARE Foundation but there are open issues on it (see comments above) that need solved by the FIWARE Foundation.

@kzangeli @flopezag taking into account the upcoming FIWARE TSC on Orion, could you provide an status update of this (or be prepared to attend to the TSC and report at the meeting), please? Thanks!

**Comment by ddigodavid** [comment]
@mapedraza @fgalan hi! do we have any news about this? How's the development going?

**Comment by fgalan** [comment]
> @mapedraza @fgalan hi! do we have any news about this? How's the development going?

We are waiting for FIWARE Foundation to address the pending comments in this PR.

File: src/app/contextBroker/contextBroker.cpp (hunk 2 above)
Comment (flopezag) [reply src/app/contextBroker/contextBroker.cpp:1246]: It should not be part of the content of this PR, If there is a previous process disabled due to the implementation of the loadFile()

**Comment by flopezag** [comment]
Ok, I was not included in the implementation process of this issue but I took a look to the code. 

Taking into account that the new approach is dynamic loading of the certification file and therefore the size of that file is irrelevant. I think that the modification of the **httpsPrepare.sh** to allow the execution of several sizes files make no sense.

**Comment by fgalan** [comment]
> Taking into account that the new approach is dynamic loading of the certification file and therefore the size of that file is irrelevant. I think that the modification of the **httpsPrepare.sh** to allow the execution of several sizes files make no sense.

Maybe not directly, but, at the end a .test to check cert sizes >2048 has to be included, and the best way to generate these cert files is the httpsPrepare.sh script, so a small modification to include the target cert size would be needed.

**Comment by flopezag** [comment]
No, I disagree, the result of that test do not introduce any relevant testing information about the code itself.

**Comment by fgalan** [comment]
> No, I disagree, the result of that test do not introduce any relevant testing information about the code itself.

A general principle in test-based developed is that test cannot assume any preconception on the implementation. This issue is about supporting certs of 2350 size so a test testing specifically that should be at least included.

...and the best way of generate a cert of 2350 cert to be used in a .test is to tune httpsPrepare.sh to to that. Other ways would involve introduce extra test tooling and that's not desirable.

**Comment by flopezag** [comment]
This suggested test taking into account the current implementation of the loadFile() should be to test the malloc function...

**Comment by fgalan** [comment]
> This suggested test taking into account the current implementation of the loadFile() should be to test the malloc function...

Note that .test test the CB as a black box, based in CLI, API calls, etc. Thus, we are not testing loadFile() function directly, but the case described by the issue associated to this PR (basically, a variant of [test/functionalTest/cases/0000_https_support/https.test](https://github.com/telefonicaid/fiware-orion/blob/master/test/functionalTest/cases/0000_https_support/https.test) with a 2350 bytes certificate).

File: src/app/contextBroker/contextBroker.cpp (hunk 2 above)
Comment (fgalan) [reply src/app/contextBroker/contextBroker.cpp:1246]: Related PR https://github.com/telefonicaid/fiware-orion/pull/3926 (where that FIXME note was included)

I think the only affected .test would be [test/functionalTest/cases/0000_https_support/https.test](https://github.com/telefonicaid/fiware-orion/blob/master/test/functionalTest/cases/0000_https_support/https.test), which already  have `VALGRIND_READY`.

Thus, NTC

**Comment by mapedraza** [comment]
Since this change is a bit more complex that increasing a constant in the program, seems reasonable (and It is always a good practice) to add test this change. Also it is described in the the [contribution guidelines](https://github.com/telefonicaid/fiware-orion/blob/0dc77185b8f98274840abc28364e2bd071c1d7e4/doc/manuals/contribution_guidelines.md#pull-request-protocol).

It is true that, at the end, this new test is almost the same that the one under `test/functionalTest/cases/0000_https_support`, but tests have to be technology agnostic, as @fgalan also explains, and if in the future, this part of the code change, both scenarios have to still be valid, and tests is the most powerful way to ensure it. That is why it is important.

The proposed changes for having the tests for this specific issue are pretty easy:
- Adding a configurable length to [`test/functionalTest/httpsPrepare.sh`](https://github.com/telefonicaid/fiware-orion/blob/master/test/functionalTest/httpsPrepare.sh)

<details>
  <summary>Proposed modifications</summary>
  
```diff
# Copyright 2014 Telefonica Investigacion y Desarrollo, S.A.U
#
# This file is part of Orion Context Broker.
#
# Orion Context Broker is free software: you can redistribute it and/or
# modify it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# Orion Context Broker is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero
# General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with Orion Context Broker. If not, see http://www.gnu.org/licenses/.
#
# For those usages not covered by this license please contact with
# iot_support at tid dot es

# -----------------------------------------------------------------------------
#
# usage
#
function usage()
{
  fileName=$(basename $0)
-  echo $fileName "[--keyFileName (name of key file)] [--certFileName (name of certificate file)]"
+  echo $fileName "[--keyFileName (name of key file)] [--certFileName (name of certificate file)] [--certFileSize (size of certificate file)]"
  echo "                [--country (country)] [--state (state)] [--city (city)] [--company (company)] [--unit (unit)] [--name (name)] [--email (email)]"
  echo
  echo "  The last seven options are input for the creation of the certificate and they all have 'decent' default values."
  exit $1
}



keyFileName="localhost.key"
certFileName="localhost.pem"
+certFileSize=2048
country="ES"
state="Madrid"
city="Madrid"
company="Telefonica"
unit="I+D"
name="localhost"
email="noone@nowhere.com"


while [ "$#" != 0 ]
do
  if   [ "$1" == "-u" ];             then usage;
  elif [ "$1" == "--keyFileName" ];  then keyFileName=$2;  shift;
  elif [ "$1" == "--certFileName" ]; then certFileName=$2; shift;
+  elif [ "$1" == "--certFileSize” ]; then certFileSize=$2; shift;
  elif [ "$1" == "--state" ];        then state=$2;        shift;
  elif [ "$1" == "--city" ];         then city=$2;         shift;
  elif [ "$1" == "--company" ];      then company=$2;      shift;
  elif [ "$1" == "--unit" ];         then unit=$2;         shift;
  elif [ "$1" == "--name" ];         then name=$2;         shift;
  elif [ "$1" == "--email" ];        then email=$2;        shift;
  else
    echo $0: bad parameter/option: "'"${1}"'";
    usage 1
  fi

  shift
done


OPTIONS="/C="$country"/ST="$state"/L="$city"/O="$company"/OU="$unit"/CN="$name"/"

-openssl genrsa -out "$keyFileName" 2048 > /dev/null 2>&1
+openssl genrsa -out "$keyFileName" $certFileSize > /dev/null 2>&1
openssl req -days 365 -out "$certFileName" -new -x509 -key "$keyFileName" -subj "$OPTIONS" > /dev/null 2>&1
```
</details>
 
- Duplicate [`test/functionalTest/cases/0000_https_support`](https://github.com/telefonicaid/fiware-orion/tree/master/test/functionalTest/cases/0000_https_support) folder (renaming it to 4014_fix_cert_length) 
- Modify `https.test` inside the new folder calling `httpsPrepare.sh` in the test with the new length 
- Add length when calling `httpsPrepare.sh` to previous tests 
   - test/functionalTest/cases/0000_https_support/https.test 
   - test/functionalTest/cases/0706_direct_https_notifications/direct_https_notifications.test
   - test/functionalTest/cases/0706_direct_https_notifications/direct_https_notifications_no_accept_selfsigned.test

**Comment by kzangeli** [comment]
If all is now good, the CNR entry is still missing. Not to be forgotten ...

**Comment by mapedraza** [comment]
> If all is now good, the CNR entry is still missing. Not to be forgotten ...

Done in https://github.com/telefonicaid/fiware-orion/pull/4014/commits/472ef5c71b146f288381566153c5937dee5d8bd9

**Comment by mapedraza** [comment]
This PR now is ready to review

**Review by fgalan — APPROVED** [review]
LGTM

**Comment by mapedraza** [comment]
After rebasing the branch `hardening/add-valgrind-to-gitactions` with changes in this PR, Valgrind tests were executed thanks to PR https://github.com/telefonicaid/fiware-orion/pull/3880. 

The actions result [here](https://github.com/telefonicaid/fiware-orion/pull/3880/checks?sha=30695a03bd822432dc32f118d11558c5e1d2e748). After detailed review done by @fgalan on Skype, I write down here the conclusion:

There are 4 tests that check that uses ` httpPrepare.sh`. but 2 of them have Valgrind already disabled (they use it just for generating a certificate for `accumulator-server.py`. There is an issue about this https://github.com/telefonicaid/fiware-orion/issues/3924). The other 2 that are running pass successfully:

- `0000_https_support`, running in batch 1
- `4012_https_cert_size`, running in batch  5

Note that errors in batch 4 are not related with this PR

**Pull request merged**